chainntnfs/height_hint_cache.go

```diff
@@ -0,0 +1,284 @@
+	// The bool returned signifies whether we have a hint cached for this
+	// transaction.
+	QueryConfirmHint(txid chainhash.Hash) (uint32, bool)
+
+	// PurgeConfirmHint removes the height hint for the transactions from
+	// the cache.
+	PurgeConfirmHint(txids ...chainhash.Hash) error
+}
+
+// heightHintCache is an implementation of the HeightHintCache interface backed
+// by a bolt DB instance where the hints will be stored.
+type heightHintCache struct {
+	db *bolt.DB
+}
+
+// A compile-time check to ensure our HeightHintCache implementation satisfies
+// the interface.
+var _ HeightHintCache = (*heightHintCache)(nil)
+
+// NewHeightHintCache returns a new height hint cache backed by a database.
+func NewHeightHintCache(dbDir string) (*heightHintCache, error) {
+	if err := os.MkdirAll(dbDir, 0700); err != nil {
+		return nil, err
+	}
+
+	db, err := bolt.Open(filepath.Join(dbDir, dbName), dbFilePermission, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &heightHintCache{db}, nil
+}
+
+// CommitSpendHint commits a height hint for the outpoints to the cache.
+func (c *heightHintCache) CommitSpendHint(heightHint uint32,
+	ops ...wire.OutPoint) error {
+
+	var (
+		outpoint bytes.Buffer
+		hint     bytes.Buffer
+	)
+
+	if err := channeldb.WriteElement(&hint, heightHint); err != nil {
+		return err
+	}
+
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		spendHints, err := tx.CreateBucketIfNotExists(spendHintBucket)
+		if err != nil {
+			return err
+		}
+
+		for _, op := range ops {
+			outpoint.Reset()
+			err := channeldb.WriteElement(&outpoint, op)
+			if err != nil {
+				return err
+			}
+
+			err = spendHints.Put(outpoint.Bytes(), hint.Bytes())
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// QuerySpendHint returns the latest height hint for an outpoint. The bool
+// returned signifies whether we have a hint cached for this outpoint.
+func (c *heightHintCache) QuerySpendHint(op wire.OutPoint) (uint32, bool) {
+	var hint uint32
+	err := c.db.View(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrSpendHintNotFound
+		}
+
+		var outpoint bytes.Buffer
+		if err := channeldb.WriteElement(&outpoint, op); err != nil {
+			return err
+		}
+
+		spendHint := spendHints.Get(outpoint.Bytes())
+		if spendHint == nil {
+			return ErrSpendHintNotFound
+		}
+
+		return channeldb.ReadElement(bytes.NewReader(spendHint), hint)
+	})
+	if err != nil {
+		return 0, false
+	}
+
+	return hint, true
+}
+
+// PurgeSpendHint removes the height hint for the outpoints from the cache.
+func (c *heightHintCache) PurgeSpendHint(ops ...wire.OutPoint) error {
+	var outpoint bytes.Buffer
+
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrSpendHintNotFound
+		}
+
+		for _, op := range ops {
+			outpoint.Reset()
+			err := channeldb.WriteElement(&outpoint, op)
```

Comment:
I think we should move `codec` into its own package, if not too intrusive change.

Reply:
Fixed.